src/PetriEngine/Colored/Reduction/RedRuleDeadTransitions.cpp

```diff
@@ -6,74 +6,91 @@
+            if (place._pre.size() > place._post.size()) continue;
 
             bool ok = true;
             // Check for producers without matching consumers first
-            for(uint prod : place._pre)
-            {
+            for (uint prod: place._pre) {
                 // Any producer without a matching consumer blocks this rule
-                Transition t = red.transitions()[prod];
+                const Transition &t = red.transitions()[prod];
                 auto in = red.getInArc(p, t);
-                if(in == t.input_arcs.end())
-                {
+                if (in == t.input_arcs.end()) {
                     ok = false;
                     break;
                 }
             }
 
-            if(!ok) continue;
+            if (!ok) continue;
 
             std::set<uint32_t> notenabled;
             // Out of the consumers, tally up those that are initially not enabled by place
             // Ensure all the enabled transitions that feed back into place are non-increasing on place.
-            for(uint cons : place._post)
-            {
-                Transition t = red.transitions()[cons];
-                auto in = red.getInArc(p, t);
-                if(in->expr->weight() <= place.marking.size())
-                {
-                    // This branch happening even once means notenabled.size() != consumers.size()
-                    auto out = red.getOutArc(t, p);
-                    // Only increasing loops are not ok
+            for (uint cons: place._post) {
+                const Transition &t = red.transitions()[cons];
+
+                const auto &in = red.getInArc(p, t);
+                const auto &out = red.getOutArc(t, p);
+
+                //Cheap check
+                if (in->expr->weight() <= place.marking.size()) {
                     if (out != t.output_arcs.end() && out->expr->weight() > in->expr->weight()) {
                         ok = false;
                         break;
                     }
                 }
-                else
-                {
+
+                //slightly more expensive check
+                uint32_t colorSize = getColorSize(t);
+                if (colorSize > 10000) continue;
+                if (markingEnablesInArc(place.marking, *in, t, partition, in->expr->getColors(red.colors()))) {
+                    if (out == t.output_arcs.end()) {
+                        continue;
+                    }
+                    // Only increasing loops are not ok
+                    auto inSet = PetriEngine::Colored::extractVarMultiset(*in->expr);
+                    auto outSet = PetriEngine::Colored::extractVarMultiset(*out->expr);
+                    if (!inSet || !outSet || (*inSet).isSubsetOrEqTo(*outSet)) {
+                        if (to_string(*out->expr) != to_string(*in->expr)) {
+                            ok = false;
+                            break;
+                        }
+                    }
```

Comment:
Since `(*inSet).isSubsetOrEqTo(*outSet)` implies `to_string(*out->expr) != to_string(*in->expr)`, the following should be slightly more performant -- and easier to read.
```suggestion
                    if ((inSet && outSet && (*inSet).isSubsetOrEqTo(*outSet)) || to_string(*out->expr) != to_string(*in->expr)) {
                        ok = false;
                        break;
                    }
```

Reply:
Edited the above to strict subset

Reply:
What if it gets split up like this?
```

auto inSet = PetriEngine::Colored::extractVarMultiset(*in->expr);
auto outSet = PetriEngine::Colored::extractVarMultiset(*out->expr);

if (!inSet || !outSet) {
    if (to_string(*out->expr) != to_string(*in->expr)) {
        ok = false;
        break;
    }
}

if ((*inSet).isSubsetOrEqTo(*outSet)) {
    ok = false;
    break;
}
```